README
  The system library is a challenge to all those using the computer to
  write their own faster and better routines or to bow to the superior
  strength and skill of a true master.
 	--http://www.inner.net/users/cmetz/program-like-a-klingon

Use diet libc to statically link programs that don't need all the bloat
from glibc.

malloc, printf and scanf contributed from Olaf Dreesen.

To compile:

	$ make

make should compile the diet libc itself.

When make is done, it will have created dietlibc.a in bin-i386 (or
bin-ppc, bin-alpha, bin-sparc, bin-ppc or bin-arm, depending on your
architecture).  In that directory you will also find a program called
"diet", which you need to copy in a directory in your $PATH:

	# install bin-i386/diet /usr/local/bin

Then you can compile programs by prepending diet to the command line,
i.e.

	$ diet gcc -s -Os -pipe -o t t.c

diet is cross-compiler friendly and can also be used like this:

	$ diet sparc-linux-gcc -o t t.c

diet will then link against dietlibc.a from bin-sparc, of course.
diet comes with a man page (diet.1), which you can copy to an
appropriate location, too:

	# cp diet.1 /usr/local/man/man1

After you compiled the diet libc successfully, I invite you to check out
the embedded utils (http://www.fefe.de/embutils/).  The embedded utils
are small replacements for common utilities like mv, chown, ls, and even
a small tar that can extract tar files.


The license for the diet libc is the GNU General Public License, version
2 (as included in the file COPYING) or later.


CHANGES
0.35:
  fix pic for x32
  SECURITY: putenv removes duplicate entries from malicious environments (Daniel Hodson)
  SECURITY: fix regex stack corruption (Thorsten Glaser)
  printf("%#.0g", 1.0) -> "1." (Christopher Wellons)
  rename WANT_DYNAMIC to WANT_CTOR, add WANT_EXCEPTIONS to dietfeatures.h
  libc can't call rand() and change the seed, so introduce rand_i (Christopher Wellons)

0.34:
  move stpcpy from libcompat to lib as it is now part of POSIX (Enrico Scholz)
  add linux/seccomp.h linux/prctl.h linux/audit.h linux/elf-em.h linux/filter.h
  SECURITY: make sure all assembler files mark the stack non-executable so
    the kernel can map the stack of the whole program as non-executable
  support compiling with clang (no cross compiling yet, and the resulting
    binary code is significantly larger than with gcc)
  implement experimental C11 threads
  fix fgets (Andreas Förster)
  add eventfd, inotify_init1, epoll_init1, accept4, recvmmsg, sendmmsg
  add dup3, fanotify_init, fanotify_mark, signalfd4, timerfd_settime
  add timerfd_gettime, setns, memfd_create, getrandom, syncfs
  add preadv, pwritev
  add getcontext, setcontext, makecontext, swapcontext (i386 and x86_64 only for now)
  add implementation of hsearch and friends
  add strverscmp (like GNU version but without special treatment for leading zeros)
  fix cuserid
  SECURITY: fix integer overflow in scandir/scandir64 if your selector
    leaves more than 1 billion filenames
  fix memory leaks in out of memory error paths in regcomp and librpc
    udp and tcp server code (found by the llvm static analyzer)
  fix returning more than one match in regexec (Stefan Wagner)
  add getentropy, lazy arc4random clone (OpenBSD)
  define uintptr_t using __UINTPTR_TYPE__ (32-bit platforms may need a make clean)
  move strnlen from libcompat into libc (it's POSIX now)
  fix dn_expand argument signedness
  add explicit_bzero, timingsafe_bcmp and timingsafe_memcmp (OpenBSD APIs useful for crypto)
  i386 switched from mmap to mmap2. glibc has been doing this for ages,
    and we now follow suit (without fallback to mmap), because
    seccomp-filter can't inspect mmap arguments, so this helps security.
  plumbing to get preliminary PIE support for x86_64
  don't cause unaligned memory accesses in stackgap for the random from auxvec (Giorgio)
  implement malloc case for realpath
  use __builtin_mul_overflow in calloc if gcc is recent enough
  cleanup s390 syscalls (Christian Seiler)
  various TLS and pthread fixes (Christian Seiler)
  add sha256 and sha512 support in crypt(3)
  promote stpncpy from libcompat to main libc as the sha512 code uses it
  add x86_64 memccpy, mempcpy
  import Debian bugfixes (Christian Seiler)
  setjmp/longjmp for powerpcspe (Roland Stigge)
  add softfpu from glibc for powerpcspe (need to pass SOFTFPU=1 to make)
  ppc64 little endian port (Christian Seiler)
  mips64 little endian port (Christian Seiler)
  aarch64 port (Christian Seiler)
  parisc fixes (Gerrit Pape, John David Anglin, Helge Deller, Christian Seiler)
  fix longjmp on mobile armv7 (Christian Seiler)
  alpha fixes (Christian Seiler, Thorsten Glaser)
  powerpc fixes (Gerrit Pape, Christian Seiler)
  sparc64 fixes (Christian Seiler)
  update setjmp/longjmp on s390x to use 64-bit ABI (Christian Seiler)
  SECURITY: x32 fixes: one of them could lead to memory corruption because struct stat was declared too short (Christian Seiler)
  add really crappy pthread_rwclock implementation for openssl 1.1
  move getrandom and getentropy declarations to <sys/random.h> to match glibc 2.25
  add secure_getenv
  mempcpy -> memcpy in sha256crypt and sha512crypt
  prototype for getpagesize was wrong (return was size_t, should be int)
  split off stackgap-g.o, don't compile valgrind stuff into stackgap.o,
    diet chooses depending on presence of -g
  add mlock2
  prefer gcc4+ __CAS implementation to our own on ARM, and fix ours (Philip Zander)
  move _POSIX_MAPPED_FILES to unistd.h
  accept "x" in fopen mode

0.33:
  first stab at getting TLS to work in actual threads
  use fs: to find current thread on x86_64
  add fallocate, posix_fallocate
  work around Ubuntu who added -fstack-protector to their gcc, which
    added stack protector code to the function from the dietlibc startup
    code that sets up the registers and segments used by the stack
    protector mechanism.
  fix sys/personality.h (Mathias Krause)
  update sys/ptrace.h (Nikola Vladov)
  fix fork+pthread closing fd 0 (Tim München)
  finally add stdout tty detection and turn off linewise buffering if !tty
  enlarge dev_t to 32-bit on i386, arm, sparc, s390 (Pascal Terjan)
  update sys/stat.h to include nsec and enlarge dev_t
  add WANT_VALGRIND_SUPPORT in dietfeatures.h.  When #defined, the
    startup code will check if the program is running under valgrind,
    and if it is, the SIMD optimized string routines will switch back to
    a slower version that does not generate valgrind false positives.
  add WANT_LLMNR in dietfeatures.h.  Now you can do "ping windowsbox"
    and it works if that box runs win7 and trusts the LAN; the down side
    is that LLMNR does not like the .local prefix, so DNS does zeroconf
    lookup for unqualified names now, too.
  add WANT_GNU_STARTUP_BLOAT in dietfeatures.h to support
    program_invocation_short_name and program_invocation_name
  add INT*_MIN, INT*_MAX, UINT*_MAX, etc to stdint.h
  remove const from stdsignal prototype to agree with standard
  only count %[] conversions in *scanf if they matched at least one char
    (Indan Zupancic)
  make sure __sc_nr_cpus does not dump core if /proc/cpuinfo is larger
    than 2048 bytes (Bela Lubkin)
  use vsyscall for time on x86_64 instead of wrapping gettimeofday (Nikola Vladov)
  CPU affinity support (Indan Zupancic)
  lots of ARM fixes (Enrico Scholz)
  make sure strftime does not write more than max bytes when writing \0
    (Tobias S. Josefowitz)
  change gettimeofday vsyscall to use the vdso

diet.c
#include <string.h>
#include <unistd.h>
#include <fcntl.h>
#include <stdio.h>
#include <stdlib.h>
#include <write12.h>

#include "dietfeatures.h"

/* goal:
 *   when invoked as
 * "diet gcc -c t.c"
 *   exec
 * "gcc -I/path/to/dietlibc/include -c t.c"
 *
 *   when invoked as
 * "diet sparc-linux-gcc -o t t.o"
 *   exec
 * "sparc-linux-gcc -nostdlib -static -o t t.o /path/to/dietlibc/bin-sparc/start.o /path/to/dietlibc/bin-sparc/dietlibc.a"
*/

static void error(const char *message) {
  __write2(message);
  exit(1);
}

static const char* Os[] = {
  "i386","-Os","-mpreferred-stack-boundary=4",
	 "-falign-functions=1","-falign-jumps=1",
	 "-falign-loops=1","-fomit-frame-pointer",0,
  "x86_64","-Os",0,
  "x32","-Os",0,
  "sparc","-Os","-mcpu=supersparc",0,
  "sparc64","-Os","-m64","-mhard-quad-float",0,
  "alpha","-Os","-fomit-frame-pointer",0,
  "arm","-Os","-fomit-frame-pointer",0,
  "aarch64","-Os","-fomit-frame-pointer",0,
  "mips","-Os","-fomit-frame-pointer","-march=mips2",0,
  "mipsel","-Os","-fomit-frame-pointer","-march=mips2",0,
  "mips64","-Os","-fomit-frame-pointer",0,
  "ppc","-Os","-fomit-frame-pointer","-mpowerpc-gpopt","-mpowerpc-gfxopt",0,
  "ppc64","-Os","-fomit-frame-pointer","-mpowerpc-gpopt","-mpowerpc-gfxopt",0,
  "ppc64le","-Os","-fomit-frame-pointer","-mpowerpc-gpopt","-mpowerpc-gfxopt",0,
  "s390","-Os","-fomit-frame-pointer",0,
  "s390x","-Os","-fomit-frame-pointer",0,
  "sh","-Os","-fomit-frame-pointer",0,
  "ia64","-Os","-fno-omit-frame-pointer",0,
  0};

static void usage(void) {
  __write2(
#ifdef __DYN_LIB
	   "dyn-"
#endif
	   "diet version " VERSION
#ifndef INSTALLVERSION
	   " (non-install version in source tree)"
#endif
	   "\n\n");
  error("usage: diet [-v] [-Os] gcc command line\n"
	"e.g.   diet -Os gcc -c t.c\n"
	"or     diet sparc-linux-gcc -o foo foo.c bar.o\n");
}

int main(int argc,char *argv[]) {
  int _link=0;
  int compile=0;
  int preprocess=0;
  int verbose=0;
  int profile=0;
  char* diethome;
  char* platform;
#ifdef __DYN_LIB
  int shared=0;
#endif
  char* shortplatform=0;
#ifdef WANT_SAFEGUARD
  char safeguard1[]="-include";
  char* safeguard2;
#endif
  const char *nostdlib="-nostdlib";
  const char *libgcc="-lgcc";
  char *libpthread="-lpthread";
  char dashL[1000];
  char dashstatic[]="-static";
  int i;
  int mangleopts=0;
  int printpath=0;
  char manglebuf[1024];
  int m,pie;

  if (!(diethome = getenv("DIETHOME")))
    diethome=DIETHOME;
#ifdef WANT_SAFEGUARD
  safeguard2=alloca(strlen(diethome)+30);
  strcpy(safeguard2, diethome);
  strcat(safeguard2, "/include/dietref.h");
#endif
  platform=alloca(strlen(diethome)+100);
  strcpy(platform,diethome);
#ifdef INSTALLVERSION
  strcat(platform,"/lib-");
#else
#ifndef __DYN_LIB
  strcat(platform,"/bin-");
#else
  strcat(platform,"/pic-");
#endif
#endif
  strcpy(dashL,"-L");

  do {
    if (!argv[1]) usage();
    if (!strcmp(argv[1],"-v")) {
      ++argv; --argc;
      verbose=1;
    } else if (!strcmp(argv[1],"-Os")) {
      ++argv; --argc;
      mangleopts=1;
    } else if (!strcmp(argv[1],"-L")) {
      ++argv; --argc;
      printpath=1;
    } else break;
  } while (1);
  {
    m=0;
    pie=0;
    for (i=1; i<argc; ++i) {
      if (!strcmp(argv[i],"-m32")) m=32; else
      if (!strcmp(argv[i],"-mx32")) m=33; else
      if (!strcmp(argv[i],"-m64")) m=64; else
      if (!strcmp(argv[i],"-pie")) pie=1; else
      if (!strcmp(argv[i],"-fpie")) pie=1; else
      if (!strcmp(argv[i],"-fno-pie")) pie=0;
    }
  }
  {
    char *cc=argv[1];
    char *tmp=strchr(cc,0)-2;
    char *tmp2,*tmp3;
    if (tmp<cc) goto donttouch;
    if (!strstr(cc,"cc") && !strstr(cc,"clang")) goto donttouch;
    if ((tmp2=strstr(cc,"linux-"))) {	/* cross compiling? */
      int len=strlen(platform);
      --tmp2;
      tmp3=strchr(cc,'-');
      if (tmp3<tmp2) tmp2=tmp3;
      if (tmp2-cc>90) error("platform name too long!\n");
      shortplatform=platform+len;
      memmove(shortplatform,argv[1],(size_t)(tmp2-cc));
      platform[tmp2-cc+len]=0;
      if (shortplatform[0]=='i' && shortplatform[2]=='8' && shortplatform[3]=='6') shortplatform[1]='3';
      if (!strncmp(shortplatform, "powerpc", 7)) {
        shortplatform[0] = shortplatform[1] = 'p'; shortplatform[2] = 'c';
        memmove(&shortplatform[3], &shortplatform[7], strlen(&shortplatform[7]) + 1);
      }
      if (!strcmp(shortplatform, "hppa"))
        strcpy(shortplatform, "parisc");
      if (!strcmp(shortplatform, "mips64el"))
        strcpy(shortplatform, "mips64");
      if (!strncmp(cc, "x86_64-linux-gnux32-", 20) || !strncmp(cc, "x86_64-pc-linux-gnux32-", 23))
        strcpy(shortplatform, "x32");
    } else {
#ifdef __sparc__
#ifdef __arch64__
      shortplatform="sparc64";
#else
      shortplatform="sparc";
#endif
#endif
#ifdef __powerpc__
      shortplatform="ppc";
#endif
#ifdef __powerpc64__
#ifdef __LITTLE_ENDIAN__
      shortplatform="ppc64le";
#else
      shortplatform="ppc64";
#endif
#endif
#ifdef __i386__
      shortplatform="i386";
#endif
#ifdef __alpha__
      shortplatform="alpha";
#endif
#ifdef __arm__
      shortplatform="arm";
#endif
#ifdef __aarch64__
      shortplatform="aarch64";
#endif
#ifdef __mips64__
      shortplatform="mips64";
#else
#ifdef __MIPSEL__
      shortplatform="mipsel";
#endif
#ifdef __MIPSEB__
      shortplatform="mips";
#endif
#endif
#ifdef __s390x__
      shortplatform="s390x";
#else
#ifdef __s390__
      shortplatform="s390";
#endif
#endif
#ifdef __sh__
      shortplatform="sh";
#endif
#ifdef __hppa__
      shortplatform="parisc";
#endif
#ifdef __x86_64__
      shortplatform=(m==32?"i386":(m==33?"x32":"x86_64"));
#endif
#ifdef __ia64__
      shortplatform="ia64";
#endif
      {
	char *tmp4=platform+strlen(platform);
	strcpy(tmp4,shortplatform);
	shortplatform=tmp4;
      }
    }
    /* MIPS needs special handling.  If argv contains -EL, change
     * platform name to mipsel */
    if (!strcmp(shortplatform,"mips")) {
      for (i=1; i<argc; ++i)
	if (!strcmp(argv[i],"-EL"))
	  strcpy(shortplatform,"mipsel");
    }
    if (printpath) {
      write(1,platform,strlen(platform));
      return 0;
    }
    strcat(dashL,platform);
    if (strcmp(tmp,"ld")) {
      char **newargv;
      char **dest;
      char *a,*b,*c;
#if defined(WANT_DYNAMIC) || defined(WANT_CTOR) || defined(WANT_EXCEPTIONS)
      char *d,*e;
#endif
#ifdef __DYN_LIB
      char *f;
#endif
      char *g=0;
/* we need to add -I... if the command line contains -c, -S or -E */
      for (i=2; i<argc; ++i) {
	if (argv[i][0]=='-') {
	  if (argv[i][1]=='M')
	    goto pp;
	  else if (argv[i][1]=='g') {
	    g=alloca(strlen(platform)+20);
	    strcpy(g,platform);
	    strcat(g,"/stackgap-g.o");
	  }
	}
	if (!strcmp(argv[i],"-pg"))
	  profile=1;
	else if (!strcmp(argv[i],"-c") || !strcmp(argv[i],"-S"))
	  compile=1;
	else if (!strcmp(argv[i],"-E"))
pp:
	  preprocess=compile=1;
      }
/* we need to add -nostdlib if we are not compiling */
      _link=!compile;
#ifdef __DYN_LIB
      if (_link) {
	for (i=2; i<argc; ++i)
	  if (!strcmp(argv[i],"-shared")) {
	    shared=1;
	    _link=0;
	  }
      }
#endif
#if 0
      for (i=2; i<argc; ++i)
	if (!strcmp(argv[i],"-o"))
	  if (!compile) _link=1;
#endif

      newargv=alloca(sizeof(char*)*(argc+100));
      a=alloca(strlen(diethome)+20);
      b=alloca(strlen(platform)+20);
      c=alloca(strlen(platform)+20);

      strcpy(a,diethome); strcat(a,"/include");
#ifndef __DYN_LIB
      strcpy(b,platform);
      if (profile)
	strcat(b,"/pstart.o");
      else
	strcat(b,pie ? "/start-pie.o" : "/start.o");
#ifdef INSTALLVERSION
      strcpy(c,platform); strcat(c,"/libc.a");
#else
      strcpy(c,platform); strcat(c,"/dietlibc.a");
#endif
#else
      strcpy(b,platform); strcat(b,"/dstart.o");
      strcpy(c,"-lc");
#endif

#ifdef WANT_DYNAMIC
      d=alloca(strlen(platform)+20);
      e=alloca(strlen(platform)+20);
#ifdef __DYN_LIB
      strcpy(d,platform);
      strcpy(e,platform);
      if (shared)
	strcat(d,"/dyn_so_start.o");
#ifdef INSTALLVERSION
      else
	strcat(d,"/dyn_dstart.o");
      strcat(e,"/dyn_dstop.o");
#else
      else
	strcat(d,"/dyn_start.o");
      strcat(e,"/dyn_stop.o");
#endif
#else
      strcpy(d,platform); strcat(d,"/dyn_start.o");
      strcpy(e,platform); strcat(e,"/dyn_stop.o");
#endif

#elif defined(WANT_CTOR) || defined(WANT_EXCEPTIONS)
      e=alloca(strlen(platform)+20);
      strcpy(e,platform);
      strcat(e,"/crtend.o");
#endif

      dest=newargv;
      *dest++=argv[1];
      if (argv[2]) {
	if (!strcmp(argv[2],"-V")) {
	  *dest++=argv[2];
	  *dest++=argv[3];
	  argv+=2;
	  argc-=2;
	} else if (!memcmp(argv[2],"-V",2)) {
	  *dest++=argv[2];
	  ++argv;
	  --argc;
	}
      }
#ifndef __DYN_LIB
      if (_link) {
	*dest++=(char*)nostdlib;
	if (pie) {
	  *dest++="-Wl,-pie";
	  *dest++="-Wl,--no-dynamic-linker";
	} else
	  *dest++=dashstatic;
	*dest++=dashL;
      }
#else
      /* avoid R_*_COPY relocations */
      *dest++="-fPIC";
      if (_link || shared) { *dest++=(char*)nostdlib; *dest++=dashL; }
#endif
#ifdef WANT_SAFEGUARD
      if (compile && !preprocess) {
	*dest++=safeguard1;
	*dest++=safeguard2;
      }
#endif
      if (_link) { *dest++=b; if (g && !pie) *dest++=g; }
#ifdef WANT_DYNAMIC
      if (_link) { *dest++=d; }
#endif
      for (i=2; i<argc; ++i) {
	if (!strcmp(argv[i],"-pthread")) {
	  *dest++="-D_REENTRANT";
	  if (_link) *dest++="-lpthread";
	  continue;
	}
#if 0
	if (mangleopts)
	  if (argv[i][0]=='-' && (argv[i][1]=='O' || argv[i][1]=='f' ||
				  (argv[i][1]=='m' && argv[i][2]!='3' && argv[i][2]!='6' && argv[i][2]!='x'))) {
	    if (strcmp(argv[i],"-fpic") && strcmp(argv[i],"-fno-pic") &&
		strcmp(argv[i],"-fpie") && strcmp(argv[i],"-fno-pie") &&
		strncmp(argv[i],"-fvisibility=",13))
	      continue;
	  }
#endif
	*dest++=argv[i];
      }
#ifndef __DYN_LIB
      if (compile || _link) {
	*dest++="-isystem";
	*dest++=a;
      }
#else
      if (compile || _link || shared) {
	*dest++="-isystem";
	*dest++=a;
      }
#endif
      *dest++="-D__dietlibc__";
      if (mangleopts) {
	const char **o=Os;

	{
	  int fd;
	  tmp=getenv("HOME");
	  if (tmp) {
	    if (strlen(tmp)+strlen(cc)<900) {
	      strcpy(manglebuf,tmp);
	      strcat(manglebuf,"/.diet/");
	      tmp=strrchr(cc,'/');
	      if (tmp) ++tmp; else tmp=cc;
	      strcat(manglebuf,tmp);
	      if ((fd=open(manglebuf,O_RDONLY))>=0) {
		int len=read(fd,manglebuf,1023);
		if (len>0) {
		  manglebuf[len]=0;
		  *dest++=manglebuf;
		  for (i=1; i<len; ++i) {
		    if (manglebuf[i]==' ' || manglebuf[i]=='\n') {
		      manglebuf[i]=0;
		      if (i+1<len)
			*dest++=manglebuf+i+1;
		    }
		  }
		  goto incorporated;
		}
	      }
	    }
	  }
	}
	for (o=Os;*o;++o) {
	  if (!strcmp(*o,shortplatform)) {
	    ++o;
	    while (*o) {
	      *dest++=(char*)*o;
	      ++o;
	    }
	    break;
	  } else
	    while (*o) ++o;
	}
      }
incorporated:
      if (_link) {
	if (profile) *dest++="-lgmon";
	*dest++=c; *dest++=(char*)libgcc;
	if (!strcmp(shortplatform,"sparc") ||
	    !strcmp(shortplatform,"sparc64") ||
	    !strncmp(shortplatform,"arm",3)) {
	  *dest++=c;
	}
      }
#if defined(WANT_DYNAMIC) || defined(WANT_CTOR) || defined(WANT_EXCEPTIONS)
      if (_link) { *dest++=e; }
#endif
#ifdef __DYN_LIB
      if (shared){ *dest++=c; }
      f=alloca(strlen(platform)+100);
      if (_link) {
	strcpy(f,"-Wl,-dynamic-linker=");
	strcat(f,platform);
//	strcat(f,"/diet-linux.so");
	strcat(f,"/libdl.so");
	*dest++=f;
      }
#endif
      *dest=0;
      if (verbose) {
	for (i=0; newargv[i]; i++) {
	  __write2(newargv[i]);
	  __write2(" ");
	}
	__write2("\n");
      }
      execvp(newargv[0],newargv);
      goto error;
    }
  }
donttouch:
  execvp(argv[1],argv+1);
error:
  error("execvp() failed!\n");
  return 1;
}
